packages/framework/immutable-merge/src/Merge.ts

```diff
@@ -51,76 +128,78 @@ export type IMergeRecursionHandler = (options: IMergeOptions, ...objs: any[]) =>
     // now assign everything to get the normal property precedence (and merge all the keys)
-    const result = Object.assign({}, ...setToMerge);
-
-    // if there is a possibility of needing the recurse, process the keys
-    for (const key in result) {
-      if (result.hasOwnProperty(key)) {
-        const originalVal = result[key];
-
-        // next options is only set if there is a possibility of recursion
-        if (nextOptions) {
-          const isObject = _isNonArrayObject(originalVal);
-          const recurseKey = recurse && recurse[key];
-          const customHandler = recurseKey && typeof recurseKey === 'function' ? recurseKey : false;
-
-          // if this key has a custom handler, or is an object eligible for recursion, recurse
-          if (customHandler || (isObject && (depth || recurseKey))) {
-            // collect all the keys, filtering out undefined values
-            const collectedObjects = setToMerge
-              .map((value: any) => {
-                return _collectVal(value && value[key], !!customHandler);
-              })
-              .filter(v => v !== undefined);
-
-            // either recurse to this function or call the custom handler if specified
-            const handler: IMergeRecursionHandler = customHandler || immutableMergeCore;
-            result[key] = handler(nextOptions, ...collectedObjects);
-            hasChanged = hasChanged || result[key] !== originalVal;
+    let result = processSingle ? undefined : Object.assign({}, ...setToMerge);
+    const processSet = result || setToMerge[0];
+
+    for (const key in processSet) {
+      if (processSet.hasOwnProperty(key)) {
+        // only process if there is potential work to do
+        if (mightRecurse) {
+          const originalVal = processSet[key];
+          const entityType = getEntityType(originalVal);
+          const handler = getHandlerForPropertyOfType(options, key, entityType);
+          if (handler !== undefined) {
+            const values = setToMerge.map(set => set[key]).filter(v => v !== undefined);
+            const updatedVal = typeof handler === 'function' ? handler(...values) : immutableMergeWorker(handler, singleMode, ...values);
+            if (updatedVal !== originalVal) {
+              result = result || Object.assign({}, ...setToMerge);
+              result[key] = updatedVal;
+            }
           }
         }
 
         // delete undefined keys from the object, otherwise there is no easy way to delete keys
-        if (result[key] === undefined) {
+        if (!processSingle && result[key] === undefined) {
           delete result[key];
         }
       }
     }
+
     // in the single processing case return the original if nothing changed, otherwise return result
-    return processSingle && !hasChanged ? setToMerge[0] : result;
+    return result || processSet;
   }
   return setToMerge.length > 0 ? setToMerge[0] : undefined;
```

Comment:
Should you warn or something if you get here. Processing a single with a set > 1 breaks the api contract right?

Reply:
Not necessarily.  The process single mode can still do multiple objects, it is just that in cases where there is a single branch it will keep processing rather than bailing early.  The signature is still ...objs: (T | undefined)[] on that function.